File: src/AutoContext.Worker.Shared/Logging/LogServerClient.cs
namespace AutoContext.Worker.Logging;

using System.IO.Pipes;
using System.Text;
using System.Text.Json;
using System.Threading.Channels;

using AutoContext.Worker.Hosting;

using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Options;

/// <summary>
/// Background pipe-client that drains worker <see cref="LogRecord"/> values
/// from a bounded channel and writes them as NDJSON over a named pipe to
/// the extension-side LogServer. When the pipe is unavailable (no
/// <c>--log-pipe</c> argument, the connect attempt fails, or the pipe
/// subsequently breaks) the client transparently falls back to writing
/// each record to stderr in a human-readable single-line format.
/// </summary>
/// <remarks>
/// The bounded channel uses <see cref="BoundedChannelFullMode.DropOldest"/>
/// — log spam can never block worker code or grow memory unbounded. A
/// single drain task owns all I/O. Failures (broken pipe, serialisation
/// errors) are intentionally swallowed: this type IS the logger of last
/// resort, so it must never throw.
/// </remarks>
internal sealed class LogServerClient : IAsyncDisposable
{
    private const int QueueCapacity = 1024;
    private const int ConnectTimeoutMs = 2000;

    private static readonly UTF8Encoding Utf8NoBom = new(encoderShouldEmitUTF8Identifier: false);

    private readonly Channel<LogRecord> _queue = Channel.CreateBounded<LogRecord>(
        new BoundedChannelOptions(QueueCapacity)
        {
            FullMode = BoundedChannelFullMode.DropOldest,
            SingleReader = true,
            SingleWriter = false,
        });

    private readonly string _pipeName;
    private readonly string _clientName;
    private readonly CancellationTokenSource _cts = new();
    private readonly Task _drainTask;

    public LogServerClient(IOptions<WorkerHostOptions> options, IHostEnvironment env)
    {
        ArgumentNullException.ThrowIfNull(options);
        ArgumentNullException.ThrowIfNull(env);

        _pipeName = options.Value.LogPipe;
        _clientName = env.ApplicationName;
        _drainTask = Task.Run(() => DrainAsync(_cts.Token));
    }

    /// <summary>
    /// Enqueues <paramref name="record"/> for off-thread delivery. Never
    /// blocks; if the queue is full the oldest record is dropped.
    /// </summary>
    public void Enqueue(LogRecord record) => _queue.Writer.TryWrite(record);

    public async ValueTask DisposeAsync()
    {
        _queue.Writer.TryComplete();

        try
        {
            await _drainTask.WaitAsync(TimeSpan.FromSeconds(2)).ConfigureAwait(false);
        }
        catch (TimeoutException)
        {
            // Drain didn't finish in time — abandon it.
        }
        catch (OperationCanceledException)
        {
            // Expected during shutdown.
        }

        await _cts.CancelAsync().ConfigureAwait(false);

        // Only dispose the CTS if the drain task has actually exited.
        // Disposing while the task is still mid-await would surface as an
        // ObjectDisposedException on the token — not caught by the drain's
        // OperationCanceledException handler, leading to an unobserved
        // task exception. The CTS is one-per-process; letting the GC
        // reclaim it is harmless.
        if (_drainTask.IsCompleted)
        {
            _cts.Dispose();
        }
    }

    private async Task DrainAsync(CancellationToken ct)
    {
        NamedPipeClientStream? stream = await TryConnectAsync(ct).ConfigureAwait(false);

        try
        {
            if (stream is not null && !await TrySendGreetingAsync(stream, ct).ConfigureAwait(false))
            {
                await stream.DisposeAsync().ConfigureAwait(false);
                stream = null;
            }

            await foreach (var record in _queue.Reader.ReadAllAsync(ct).ConfigureAwait(false))
            {
                if (stream is not null)
                {
                    if (await TryWritePipeAsync(stream, record, ct).ConfigureAwait(false))
                    {
                        continue;
                    }

                    await stream.DisposeAsync().ConfigureAwait(false);
                    stream = null;
                }

                WriteStderr(record);
            }
        }
        catch (OperationCanceledException)
        {
            // Shutdown — fall through.
        }
        finally
        {
            if (stream is not null)
            {
                await stream.DisposeAsync().ConfigureAwait(false);
            }
        }
    }

    private async Task<NamedPipeClientStream?> TryConnectAsync(CancellationToken ct)
    {
        if (string.IsNullOrEmpty(_pipeName))
        {
            return null;
        }

        var pipe = new NamedPipeClientStream(
            serverName: ".",
            pipeName: _pipeName,
            direction: PipeDirection.Out,
            options: PipeOptions.Asynchronous);

        try
        {
            await pipe.ConnectAsync(ConnectTimeoutMs, ct).ConfigureAwait(false);
            return pipe;
        }
        catch (Exception ex) when (ex is TimeoutException or IOException or UnauthorizedAccessException or OperationCanceledException)
        {
            await pipe.DisposeAsync().ConfigureAwait(false);
            return null;
        }
    }

    private async Task<bool> TrySendGreetingAsync(Stream stream, CancellationToken ct)
    {
        try
        {
            var json = JsonSerializer.Serialize(
                new LogGreetingWire(_clientName),
                LogServerJsonContext.Default.LogGreetingWire);
            var bytes = Utf8NoBom.GetBytes(json + "\n");
            await stream.WriteAsync(bytes, ct).ConfigureAwait(false);
            await stream.FlushAsync(ct).ConfigureAwait(false);
            return true;
        }
        catch (Exception ex) when (ex is IOException or ObjectDisposedException)
        {
            return false;
        }
    }

    private static async Task<bool> TryWritePipeAsync(Stream stream, LogRecord record, CancellationToken ct)
    {
        try
        {
            var wire = new LogRecordWire(
                Category: record.Category,
                Level: record.Level.ToString(),
                Message: record.Message,
                Exception: record.Exception?.ToString(),
                CorrelationId: record.CorrelationId);
            var json = JsonSerializer.Serialize(wire, LogServerJsonContext.Default.LogRecordWire);
            var bytes = Utf8NoBom.GetBytes(json + "\n");
            await stream.WriteAsync(bytes, ct).ConfigureAwait(false);
            await stream.FlushAsync(ct).ConfigureAwait(false);
            return true;
        }
        catch (Exception ex) when (ex is IOException or ObjectDisposedException)
        {
            return false;
        }
    }

    private static void WriteStderr(LogRecord record)
    {
        try
        {
            var prefix = record.CorrelationId is null ? string.Empty : $"[{record.CorrelationId}] ";
            var line = $"{prefix}{record.Level}: {record.Category}: {record.Message}";
            Console.Error.WriteLine(line);
            if (record.Exception is not null)
            {
                Console.Error.WriteLine(record.Exception);
            }
        }
        catch (IOException)
        {
            // stderr is gone — nothing we can do.
        }
    }
}

File: src/AutoContext.Worker.Web/src/logging/log-server-client.ts
import { connect, type Socket } from 'node:net';
import type { LogGreetingWire, LogRecord, LogRecordWire } from '#types/log-record.js';
import type { LogSink } from '#types/log-sink.js';

/**
 * Background pipe-client that drains worker {@link LogRecord} values
 * from a bounded queue and writes them as NDJSON over a named pipe to
 * the extension-side LogServer. When the pipe is unavailable (no
 * `--log-pipe` argument, the connect attempt fails, or the pipe
 * subsequently breaks) the client transparently falls back to writing
 * each record to stderr in a human-readable single-line format.
 *
 * The bounded queue uses drop-oldest semantics — log spam can never
 * block worker code or grow memory unbounded. A single drain task
 * owns all I/O. Failures (broken pipe, serialisation errors) are
 * intentionally swallowed: this type IS the logger of last resort,
 * so it must never throw.
 *
 * TypeScript counterpart of `LogServerClient` in
 * `AutoContext.Worker.Shared`.
 */
export class LogServerClient implements LogSink {
    private static readonly QUEUE_CAPACITY = 1024;
    private static readonly CONNECT_TIMEOUT_MS = 2000;

    private readonly pipeName: string;
    private readonly clientName: string;
    private readonly queue: LogRecord[] = [];
    private readonly waiters: Array<() => void> = [];
    private completed = false;
    private socket: Socket | undefined;
    private readonly drainTask: Promise<void>;
    private disposeTask: Promise<void> | undefined;

    constructor(pipeName: string, clientName: string) {
        if (clientName.trim() === '') {
            throw new Error('clientName must be a non-empty string.');
        }
        this.pipeName = pipeName;
        this.clientName = clientName;
        this.drainTask = this.drain();
    }

    /**
     * Enqueues `record` for off-thread delivery. Never blocks; if the
     * queue is full the oldest record is dropped.
     */
    enqueue(record: LogRecord): void {
        if (this.completed) {
            return;
        }
        if (this.queue.length >= LogServerClient.QUEUE_CAPACITY) {
            this.queue.shift();
        }
        this.queue.push(record);
        const waiter = this.waiters.shift();
        if (waiter !== undefined) {
            waiter();
        }
    }

    /**
     * Stops accepting new records, waits up to two seconds for the
     * drain loop to flush, then tears the socket down. Safe to call
     * multiple times — concurrent callers share a single completion
     * promise.
     */
    async dispose(): Promise<void> {
        this.disposeTask ??= this.disposeCore();
        return this.disposeTask;
    }

    private async disposeCore(): Promise<void> {
        this.completed = true;
        // Wake any pending wait so the drain loop can observe completion.
        const waiters = this.waiters.splice(0);
        for (const waiter of waiters) {
            waiter();
        }

        let timer: NodeJS.Timeout | undefined;
        const timeout = new Promise<void>((resolve) => {
            timer = setTimeout(resolve, 2000);
            timer.unref();
        });
        try {
            await Promise.race([this.drainTask, timeout]);
        } finally {
            if (timer !== undefined) {
                clearTimeout(timer);
            }
        }

        const socket = this.socket;
        this.socket = undefined;
        if (socket !== undefined) {
            socket.destroy();
        }
    }

    private async drain(): Promise<void> {
        let socket = await this.tryConnect();
        if (socket !== null && !(await this.trySendGreeting(socket))) {
            socket.destroy();
            socket = null;
        }
        this.socket = socket ?? undefined;

        try {
            while (true) {
                const record = await this.dequeue();
                if (record === undefined) {
                    return;
                }
                if (this.socket !== undefined) {
                    if (await this.tryWritePipe(this.socket, record)) {
                        continue;
                    }
                    this.socket.destroy();
                    this.socket = undefined;
                }
                LogServerClient.writeStderr(record);
            }
        }
        finally {
            const s = this.socket;
            this.socket = undefined;
            if (s !== undefined) {
                s.destroy();
            }
        }
    }

    private dequeue(): Promise<LogRecord | undefined> {
        if (this.queue.length > 0) {
            return Promise.resolve(this.queue.shift());
        }
        if (this.completed) {
            return Promise.resolve(undefined);
        }
        return new Promise<LogRecord | undefined>((resolve) => {
            this.waiters.push(() => {
                if (this.queue.length > 0) {
                    resolve(this.queue.shift());
                    return;
                }
                resolve(undefined);
            });
        });
    }

    private tryConnect(): Promise<Socket | null> {
        if (this.pipeName.trim() === '') {
            return Promise.resolve(null);
        }
        const pipePath = process.platform === 'win32'
            ? `\\\\.\\pipe\\${this.pipeName}`
            : `/tmp/CoreFxPipe_${this.pipeName}`;

        return new Promise<Socket | null>((resolve) => {
            const sock = connect(pipePath);
            let settled = false;
            const settle = (value: Socket | null): void => {
                if (settled) {
                    return;
                }
                settled = true;
                clearTimeout(timer);
                sock.removeListener('connect', onConnect);
                sock.removeListener('error', onError);
                if (value === null) {
                    sock.destroy();
                }
                resolve(value);
            };
            const onConnect = (): void => settle(sock);
            const onError = (): void => settle(null);
            const timer = setTimeout(() => settle(null), LogServerClient.CONNECT_TIMEOUT_MS);
            timer.unref();
            sock.once('connect', onConnect);
            sock.once('error', onError);
        });
    }

    private trySendGreeting(socket: Socket): Promise<boolean> {
        const greeting: LogGreetingWire = { clientName: this.clientName };
        return LogServerClient.writeLine(socket, JSON.stringify(greeting));
    }

    private tryWritePipe(socket: Socket, record: LogRecord): Promise<boolean> {
        const wire: LogRecordWire = {
            category: record.category,
            level: record.level,
            message: record.message,
            ...(record.exception !== undefined ? { exception: record.exception } : {}),
            ...(record.correlationId !== undefined ? { correlationId: record.correlationId } : {}),
        };
        let json: string;
        try {
            json = JSON.stringify(wire);
        }
        catch {
            // Returning false routes the record through the stderr
            // fallback rather than silently losing it. The wire shape
            // contains only strings/undefined so this is essentially
            // unreachable today — the guard is here to keep the drain
            // loop honest if the type contract is ever violated.
            return Promise.resolve(false);
        }
        return LogServerClient.writeLine(socket, json);
    }

    private static writeLine(socket: Socket, json: string): Promise<boolean> {
        return new Promise<boolean>((resolve) => {
            if (socket.destroyed || !socket.writable) {
                resolve(false);
                return;
            }
            socket.write(json + '\n', (err) => {
                resolve(err === null || err === undefined);
            });
        });
    }

    private static writeStderr(record: LogRecord): void {
        try {
            const prefix = record.correlationId === undefined ? '' : `[${record.correlationId}] `;
            process.stderr.write(`${prefix}${record.level}: ${record.category}: ${record.message}\n`);
            if (record.exception !== undefined) {
                process.stderr.write(`${record.exception}\n`);
            }
        }
        catch {
            // stderr is gone — nothing we can do.
        }
    }
}
